### DISCUSSION TITLE
Thermostatic control

### DISCUSSION THREAD

**Discussion by OceanOv** [Rem0o/FanControl.Releases#483] [Ideas]
Hello!
Perhaps I can offer you one more cooler control logic, "thermostatic" (as in OCTune Excavator https://github.com/nicehash/NiceHashQuickMiner/wiki/OCTune).
The meaning of the logic is something like this:
1. Set the target temperature of the device (for example, the temperature of the temperature sensor of the water block or the motherboard, around 45c)
1. At the start: turn on the coolers at maximum speed.
2. As soon as the temperature drops below 45c (ie 44c), for example, the coolers begin to slow down very smoothly.
3. At the moment of slowdown, when the values ​​start jumping from 44 to 45 and back, the cooler speed will be considered "optimal".
4. But, for example, as soon as the temperature of 45 becomes stable, it will be necessary to start to increase speed very smoothly, and as soon as the values ​​begin to jump again from 44 to 45 and back, these values ​​can be considered "optimal".
And vice versa, if the temperature suddenly settled at 44 or even dropped to 43, then you need to slow down the speed, smoothly, to find the "optimal" temperature.
The user, ideally, should be given the parameters of response speed, and delays.
This would reduce the noise and wear of coolers to a minimum in conditions of changing ambient temperatures
I decided to share an idea with you, it would be very cool if it could be implemented.
Thx you for your work!

**Comment by Rem0o** [comment]
Did you try the Auto fan curve? It kind of behave like this. Check it out.

**Reply by OceanOv** [reply]
I studied it again, very carefully. No it's not that. It's like a curve that automatically calculates RPM within a given range.
It is not self-adjusting to a certain temperature (depending on the dynamics). Only work with extremes.
It turns out that the cooler constantly throws up and down.

**Reply by Rem0o** [reply]
Calculate RPM? No.

Here how it works:

It will increase or decrease % to try and keep the load temperature.

If you set 50 deg as load temp, if the temp goes over that, it will increase the % until it goes back to 50 or lower.   If it goes lower than 50, it will decrease the % slowly to find an equilibrium.

**Reply by OceanOv** [reply]
My target temperature is 44c (petg tubes start to slowly deform at temperatures above 45c on the motherboard (actual water temperature in the tubes is 40c).
Motherboard sensor values ​​are limited to an integer, without tenths (if there were tenths, then of course it would be easier to set up, and any curve).
It would be ideal to capture fluctuations at 44c between 43 and 45, gradually adding or decreasing cooler speed that touches these values. On average, I have an increase of 1c during idling coolers occurs within 10 seconds. under maximum load.
Now the coolers just jump back and forth from minimum to maximum (without choosing the optimal RPM). Perhaps I misconfigured something?
The values ​​in the program are as follows:
idle temp: 39c
load temp: 44c
min fan speed 36%
max fan speed 100%
step 2%
deadband 3%
response time 2sec
![изображение](https://user-images.githubusercontent.com/92110371/150396901-9f5e1fec-d1f8-4429-93be-e3de7e845baa.png)

**Comment by Rem0o** [comment]
Hum, I don't think this function will work well with such a small range (39 to 44). As you said, values are integers, so we got almost no play. 

I would say, for testing purposes, try and set your idle temp way lower, like 30. Set deadband to 1 deg, step to 1%.
If you get erratic behavior ( up down up down), decrease step ( you can do decimals) and increase response time. Play with these values.

The deadband will set a "no change zone" from 43 to 44 deg. 

Above 44, it should start increasing.
At 44 and 43, no change. Below 43, values will drop.
If the temp is stable between 43 and 44, there should be no change in speed, as it found an equilibrium.

**Reply by OceanOv** [reply]
The situation is the same ¯\_(ツ)_/¯
The deadband does not allow values less than 2c.
Maybe add a new thermostat function?)
Before that, I just used the curve, looking at the changes in the hwinfo charts, I chose the average between the extreme values. Therefore, now more values are built with reference to specific temperatures (which sometimes float), which will not be very convenient when summer comes and you have to build new values. The temperature in my room now ranges from 23c to 28c, and in summer it can reach 32c.
Maybe do you have any ideas on how best to build this logic? But I think a thermostat feature would be a very nice addition to your software.
![изображение](https://user-images.githubusercontent.com/92110371/150404427-5ffeb4a3-b74b-45f4-bf28-e13d9f33dfc0.png)

**Reply by Rem0o** [reply]
It "is" my best shot at a "thermostat" feature. What I can do is to try to improve it.

Keep in mind your use case is very very (very) niche.  Most people track cpu and gpu temps, which are not as stable. I've built the feature with those in mind.

However, we can try and make it work.

First thing, the deadband should allow a value of 1, I'll change that in the next update.
After that, I don't think the settings you are using are optimal yet.

What we can see from your graph, is that the function is unstable. That's because the ramp up and down is still too fast.
We can see that it starts to increase when the temp is rising and reaches 44 deg, which is expected. Then it maxes out quickly after a few seconds ( what's the scale of your graph in the x-axis ? ).

Try and set something wayyyyy slow, like 0.1% step, and set your response time to 1 sec, and see how it goes.

Something like this: 
![image](https://user-images.githubusercontent.com/14118956/150415575-30a6e607-03ed-4420-abf5-81ce55945132.png)

**Reply by OceanOv** [reply]
Just experimented with such values, it seems to work. It would be interesting, of course, to try to upgrade from the dead zone to 1c.
Thank you so much for your hard work, looking forward to the update!
![изображение](https://user-images.githubusercontent.com/92110371/150416865-cdb2fc1d-c198-429f-b236-4fa2c97f2a25.png)

**Reply by OceanOv** [reply]
> 

In the new version, everything works even better (deadband values 1). This is great, thanks!
![изображение](https://user-images.githubusercontent.com/92110371/151461508-d81bca9e-d5bd-452c-9543-99f9c7de155e.png)

**Comment by OceanOv** [comment]
I found one strong bug in autofan. When you restart the application, it starts from some average values, and does not select speeds from the very beginning. It turns out that it cannot even reach the "working temperature" and supercools the system.

**Reply by OceanOv** [reply]
[userConfig.json.txt](https://github.com/Rem0o/FanControl.Releases/files/8104202/userConfig.json.txt)
